fendermint/vm/interpreter/src/fvm/state/snapshot.rs

```diff
@@ -0,0 +1,357 @@
+        todo!()
+    }
+
+    fn car_header_roots(&self) -> anyhow::Result<Vec<Cid>> {
+        match self {
+            Snapshot::V1(s) => s.car_header_roots(),
+        }
+    }
+}
+
+pub struct V1Snapshot<DB> {
+    /// The root cid of the state tree
+    state_tree_root: Cid,
+    /// The state tree of the current blockchain
+    state_tree: StateTree<ReadOnlyBlockstore<DB>>,
+    /// The latest block height with state params serialized to bytes
+    block_state_params_bytes: Vec<u8>,
+    /// Block state params cid derived from block state params bytes
+    block_state_params_cid: Cid,
+}
+
+impl<DB> V1Snapshot<DB>
+where
+    DB: Blockstore + Clone + 'static + Send,
+{
+    /// Creates a new V2Snapshot struct. Caller ensure store
+    pub fn new(
+        store: DB,
+        state_params: FvmStateParams,
+        block_height: BlockHeight,
+    ) -> anyhow::Result<Self> {
+        let state_tree =
+            StateTree::new_from_root(ReadOnlyBlockstore::new(store), &state_params.state_root)?;
+        let state_tree_root = state_params.state_root;
+
+        let block_state_params_bytes = fvm_ipld_encoding::to_vec(&(state_params, block_height))?;
+        let block_state_params_cid =
+            Cid::new_v1(DAG_CBOR, Code::Blake2b256.digest(&block_state_params_bytes));
+
+        Ok(Self {
+            state_tree_root,
+            state_tree,
+            block_state_params_bytes,
+            block_state_params_cid,
+        })
+    }
+
+    async fn read_car(file: tokio::fs::File, store: DB) -> anyhow::Result<Self> {
+        let roots = load_car_unchecked(&store, file.compat()).await?;
+
+        if roots.len() != 2 {
+            return Err(anyhow!("invalid v1 snapshot, should have 2 root cids"));
+        }
+
+        let state_tree_root = roots[0];
+        let block_state_params_cid = roots[1];
+
+        let block_state_params_bytes = if let Some(bytes) = store.get(&block_state_params_cid)? {
+            bytes
+        } else {
+            return Err(anyhow!(
+                "invalid v1 snapshot, block state params cid not found: {}",
+                block_state_params_cid
+            ));
+        };
+
+        Ok(Self {
+            state_tree_root,
+            state_tree: StateTree::new_from_root(ReadOnlyBlockstore::new(store), &state_tree_root)?,
+            block_state_params_bytes,
+            block_state_params_cid,
+        })
+    }
+
+    /// For V1 snapshot, we are putting two components into the CAR file: the state tree and latest
+    /// state params. One header is for the state tree root. The other header root cid is for the
+    /// serialized block state params.
+    fn car_header_roots(&self) -> anyhow::Result<Vec<Cid>> {
+        Ok(vec![self.state_tree_root, self.block_state_params_cid])
+    }
+
+    pub fn init_chain(&self) -> anyhow::Result<()> {
+        todo!()
+    }
+
+    async fn write_car(self, car: CarHeader, write: tokio::fs::File) -> anyhow::Result<()> {
```

Comment:
Why create the `CarHeader` here, and then the `car_header_roots` function is not needed, you can delay serializing the `state_params` to a `Block` to get the CID up to this point, and not need to store bytes and CID in the struct?

Reply:
I updated slightly to use stream instead. But originally I was thinking the caller, i.e. `Snapshot` should control the root cids.